# Olist customer analysis with SQL

This notebook uses most of tables in the Olist database from `customers` table to `orders`, `order_items`, `products` and `sellers`

The flow of analysis would be: 
- Time series analysis with metrics such as customer growth rate and repeat customer rate
- Where are Olist's customers? What insights can we get from the location of customers?

## Connect and load in the database

In [1]:
%load_ext sql
import os

In [2]:
%reload_ext sql

In [3]:
host = "localhost"
database = "olist"
user = "postgres"
password = "1111"
connection_string = f"postgresql://{user}:{password}@{host}/{database}"
%sql $connection_string

## Time series analysis

### Customer growth rate

Before calculating the growth rate, let's see the number of customers each year.

In [61]:
%%sql
SELECT EXTRACT(YEAR FROM order_purchase) AS year,
        COUNT(DISTINCT customer_unique_id) AS no_of_customers
FROM orders
    JOIN customers USING (customer_id)
GROUP BY year
ORDER BY year

 * postgresql://postgres:***@localhost/olist
3 rows affected.


[(Decimal('2016'), 326), (Decimal('2017'), 43713), (Decimal('2018'), 52749)]

Starting with only 326 customers in 2016, the number of customers increased tremendously in 2017 and 2018. <br>
To be specific, we will see the breakdown of number of customers per each month:

In [47]:
%%sql
SELECT EXTRACT(YEAR FROM order_purchase) AS year,
        EXTRACT(MONTH FROM order_purchase) AS month,
        COUNT(DISTINCT customer_unique_id) AS no_of_customers
FROM orders
    JOIN customers USING (customer_id)
GROUP BY year, month
ORDER BY year, month

 * postgresql://postgres:***@localhost/olist
25 rows affected.


[(Decimal('2016'), Decimal('9'), 4),
 (Decimal('2016'), Decimal('10'), 321),
 (Decimal('2016'), Decimal('12'), 1),
 (Decimal('2017'), Decimal('1'), 765),
 (Decimal('2017'), Decimal('2'), 1755),
 (Decimal('2017'), Decimal('3'), 2642),
 (Decimal('2017'), Decimal('4'), 2372),
 (Decimal('2017'), Decimal('5'), 3625),
 (Decimal('2017'), Decimal('6'), 3180),
 (Decimal('2017'), Decimal('7'), 3947),
 (Decimal('2017'), Decimal('8'), 4246),
 (Decimal('2017'), Decimal('9'), 4212),
 (Decimal('2017'), Decimal('10'), 4561),
 (Decimal('2017'), Decimal('11'), 7430),
 (Decimal('2017'), Decimal('12'), 5603),
 (Decimal('2018'), Decimal('1'), 7166),
 (Decimal('2018'), Decimal('2'), 6569),
 (Decimal('2018'), Decimal('3'), 7115),
 (Decimal('2018'), Decimal('4'), 6882),
 (Decimal('2018'), Decimal('5'), 6814),
 (Decimal('2018'), Decimal('6'), 6128),
 (Decimal('2018'), Decimal('7'), 6230),
 (Decimal('2018'), Decimal('8'), 6460),
 (Decimal('2018'), Decimal('9'), 14),
 (Decimal('2018'), Decimal('10'), 4)]

There is one missing month in November 2016. As being explained by the author of this dataset, it is because they had to pause the store for the update of [a new version](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce/discussion/69728). 

Now, we are all good to go to calculate the customer growth rate. First, let's see the increasing of customer rate each year:

In [94]:
%%sql

WITH tmp AS 
(
    SELECT EXTRACT(YEAR FROM order_purchase) AS year,
            COUNT(DISTINCT customer_unique_id) AS no_of_customers
    FROM customers
        JOIN orders USING (customer_id)
    GROUP BY year
    ORDER BY year
)

SELECT *,
    ROUND(CAST(CAST((no_of_customers - LAG(no_of_customers) OVER (ORDER BY year)) AS REAL) /
                            (LAG(no_of_customers) OVER (ORDER BY year)) * 100 AS NUMERIC), 2) 
                                AS growth_rate
FROM tmp

 * postgresql://postgres:***@localhost/olist
3 rows affected.


[(Decimal('2016'), 326, None),
 (Decimal('2017'), 43713, Decimal('13308.90')),
 (Decimal('2018'), 52749, Decimal('20.67'))]

The idea here is to take the number of customers in following year minus with the previous year and still divide it to get the growth rate. Based on the above table, the growth rate in 2017 increased dramatically to 13,308% while there is just a relative increase in 2018 with 20.67%

Here, CTE and Window Function have been used for the calculation: <br>
- CTE is used to create a temporary table of extracting year with number of customers
- Window Function, specifically LAG() function is used to take the value of previous year in order to to calculate the growth rate


In [93]:
%%sql

WITH tmp AS 
(
    SELECT EXTRACT(YEAR FROM order_purchase) AS year,
            EXTRACT(MONTH FROM order_purchase) AS month, 
            COUNT(DISTINCT customer_unique_id) AS no_of_customers
    FROM customers
        JOIN orders USING (customer_id)
    GROUP BY year, month
    ORDER BY year, month
)

SELECT *,
    ROUND(CAST(CAST((no_of_customers - LAG(no_of_customers) OVER (ORDER BY year, month)) AS REAL) / 
                            (LAG(no_of_customers) OVER (ORDER BY year, month)) * 100 AS NUMERIC),2) 
                                AS growth_rate
FROM tmp

 * postgresql://postgres:***@localhost/olist
25 rows affected.


[(Decimal('2016'), Decimal('9'), 4, None),
 (Decimal('2016'), Decimal('10'), 321, Decimal('7925.00')),
 (Decimal('2016'), Decimal('12'), 1, Decimal('-99.69')),
 (Decimal('2017'), Decimal('1'), 765, Decimal('76400.00')),
 (Decimal('2017'), Decimal('2'), 1755, Decimal('129.41')),
 (Decimal('2017'), Decimal('3'), 2642, Decimal('50.54')),
 (Decimal('2017'), Decimal('4'), 2372, Decimal('-10.22')),
 (Decimal('2017'), Decimal('5'), 3625, Decimal('52.82')),
 (Decimal('2017'), Decimal('6'), 3180, Decimal('-12.28')),
 (Decimal('2017'), Decimal('7'), 3947, Decimal('24.12')),
 (Decimal('2017'), Decimal('8'), 4246, Decimal('7.58')),
 (Decimal('2017'), Decimal('9'), 4212, Decimal('-0.80')),
 (Decimal('2017'), Decimal('10'), 4561, Decimal('8.29')),
 (Decimal('2017'), Decimal('11'), 7430, Decimal('62.90')),
 (Decimal('2017'), Decimal('12'), 5603, Decimal('-24.59')),
 (Decimal('2018'), Decimal('1'), 7166, Decimal('27.90')),
 (Decimal('2018'), Decimal('2'), 6569, Decimal('-8.33')),
 (Decimal('2018'), Decimal('3'), 7115, Decimal('8.31')),
 (Decimal('2018'), Decimal('4'), 6882, Decimal('-3.27')),
 (Decimal('2018'), Decimal('5'), 6814, Decimal('-0.99')),
 (Decimal('2018'), Decimal('6'), 6128, Decimal('-10.07')),
 (Decimal('2018'), Decimal('7'), 6230, Decimal('1.66')),
 (Decimal('2018'), Decimal('8'), 6460, Decimal('3.69')),
 (Decimal('2018'), Decimal('9'), 14, Decimal('-99.78')),
 (Decimal('2018'), Decimal('10'), 4, Decimal('-71.43'))]

Sharing the same logic with yearly growth rate, we will apply the same calculation logic with CTE and Window Function but in monthly base. <br>
We will only observe the period between January 2017 and August 2018 because the remaining time seems unreliable as the change varied in large magnitude. <br>
2017 is a successful year to Olist with a stable growth while the growth was fading away in 2018.

### Repeat customer rate

In order to classify whether one customer is repeatable or not, we need to set the logic here with the help of Window Function. <br>
- The customer (representative with one unique ID) who purchased only once (as there are no other row_number for that customer but 1) would be assigned as Total
- Customers when there are more than 1 in row_number would be assigned as Return

In [222]:
%%sql
SELECT CASE ROW_NUMBER() OVER (PARTITION BY customer_unique_id ORDER BY order_purchase)
        WHEN 1 THEN 'Total'
        ELSE 'Return' 
            END AS customer_type,
        customer_unique_id,
        EXTRACT(YEAR FROM order_purchase) AS year,
        EXTRACT(MONTH FROM order_purchase) AS month

FROM customers 
    JOIN orders USING (customer_id)

ORDER BY customer_unique_id, order_purchase

LIMIT 5

 * postgresql://postgres:***@localhost/olist
5 rows affected.


[('Total', '0000366f3b9a7992bf8c76cfdf3221e2', Decimal('2018'), Decimal('5')),
 ('Total', '0000b849f77a49e4a4ce2b2a4ca5be3f', Decimal('2018'), Decimal('5')),
 ('Total', '0000f46a3911fa3c0805444483337064', Decimal('2017'), Decimal('3')),
 ('Total', '0000f6ccb0745a6a4b88665a16c9f078', Decimal('2017'), Decimal('10')),
 ('Total', '0004aac84e0df4da2b147fca70cf8255', Decimal('2017'), Decimal('11'))]

The above result would be stored in temporary table thanks to CTE. Based on that, we now can count how many repeat customers yearly.

In [233]:
%%sql

WITH tmp AS
(
    SELECT CASE ROW_NUMBER() OVER (PARTITION BY customer_unique_id ORDER BY order_purchase)
            WHEN 1 THEN 'Total'
            ELSE 'Return'
                END AS customer_type,
            customer_unique_id,
            EXTRACT(YEAR FROM order_purchase) AS year
    FROM customers
        JOIN orders USING (customer_id)
    ORDER BY customer_unique_id, order_purchase
)

SELECT year,
        COUNT(DISTINCT customer_unique_id) AS repeat_customers
FROM tmp
WHERE customer_type = 'Return'
GROUP BY year
ORDER BY year

 * postgresql://postgres:***@localhost/olist
3 rows affected.


[(Decimal('2016'), 3), (Decimal('2017'), 1261), (Decimal('2018'), 1799)]

The number of repeat customers looks good with yearly increasing in both 2017 and 2018. However, we still want to know the growth rate and the share rate of repeat customers in total number of customers.

In [248]:
%%sql

WITH tmp AS
(
    SELECT CASE ROW_NUMBER() OVER (PARTITION BY customer_unique_id ORDER BY order_purchase)
            WHEN 1 THEN 'Total'
            ELSE 'Return'
                END AS customer_type,
            customer_unique_id,
            EXTRACT(YEAR FROM order_purchase) AS year
    FROM customers
        JOIN orders USING (customer_id)
    ORDER BY customer_unique_id, order_purchase
),

repeat AS 
(
    SELECT year, 
            COUNT(DISTINCT customer_unique_id) AS repeat_customers
    FROM tmp
    WHERE customer_type = 'Return'
    GROUP BY year
    ORDER BY year

),

total AS
(
    SELECT EXTRACT(YEAR FROM order_purchase) AS year,
            COUNT(DISTINCT customer_unique_id) AS no_of_customers
    FROM customers 
        JOIN orders USING (customer_id)
    GROUP BY year
    ORDER BY year
)

SELECT year, repeat_customers,
    ROUND(CAST(CAST(repeat_customers - LAG(repeat_customers) OVER (ORDER BY year) AS REAL) / 
                        LAG(repeat_customers) OVER (ORDER BY year) * 100 AS NUMERIC), 2) AS repeat_growth_rate,
    ROUND(CAST(CAST(repeat_customers AS REAL) / no_of_customers * 100 AS NUMERIC), 2) AS repeat_share
    
FROM repeat
    JOIN total USING (year)

 * postgresql://postgres:***@localhost/olist
3 rows affected.


[(Decimal('2016'), 3, None, Decimal('0.92')),
 (Decimal('2017'), 1261, Decimal('41933.33'), Decimal('2.88')),
 (Decimal('2018'), 1799, Decimal('42.66'), Decimal('3.41'))]

The growth rate of repeat customers followed the trend of customer growth rate but did better in the period between 2017 and 2018 when the rate increased around 43% comparing with only 21% of overall customer growth rate. <br>
Meanwhile, the repeat customer accounted for a very small ratio in the total customer each year.

In [247]:
%%sql

WITH tmp AS
(
    SELECT CASE ROW_NUMBER() OVER (PARTITION BY customer_unique_id ORDER BY order_purchase)
            WHEN 1 THEN 'Total'
            ELSE 'Return'
                END AS customer_type,
            customer_unique_id,
            EXTRACT(YEAR FROM order_purchase) AS year,
            EXTRACT(MONTH FROM order_purchase) AS month
    FROM customers
        JOIN orders USING (customer_id)
    ORDER BY customer_unique_id, order_purchase
),

repeat AS 
(
    SELECT year, month,
            COUNT(DISTINCT customer_unique_id) AS repeat_customers
    FROM tmp
    WHERE customer_type = 'Return'
    GROUP BY year, month
    ORDER BY year, month

),

total AS
(
    SELECT EXTRACT(YEAR FROM order_purchase) AS year,
            EXTRACT(MONTH FROM order_purchase) AS month,
            COUNT(DISTINCT customer_unique_id) AS no_of_customers
    FROM customers 
        JOIN orders USING (customer_id)
    GROUP BY year, month
    ORDER BY year, month
)

SELECT year, month, repeat_customers,
    ROUND(CAST(CAST(repeat_customers - LAG(repeat_customers) OVER (ORDER BY year) AS REAL) / 
                        LAG(repeat_customers) OVER (ORDER BY year) * 100 AS NUMERIC), 2) AS repeat_growth_rate,
    ROUND(CAST(CAST(repeat_customers AS REAL) / no_of_customers * 100 AS NUMERIC), 2) AS repeat_share
    
FROM repeat
    JOIN total USING (year, month)

 * postgresql://postgres:***@localhost/olist
23 rows affected.


[(Decimal('2016'), Decimal('10'), 3, None, Decimal('0.93')),
 (Decimal('2017'), Decimal('1'), 27, Decimal('800.00'), Decimal('3.53')),
 (Decimal('2017'), Decimal('2'), 25, Decimal('-7.41'), Decimal('1.42')),
 (Decimal('2017'), Decimal('3'), 45, Decimal('80.00'), Decimal('1.70')),
 (Decimal('2017'), Decimal('4'), 47, Decimal('4.44'), Decimal('1.98')),
 (Decimal('2017'), Decimal('5'), 101, Decimal('114.89'), Decimal('2.79')),
 (Decimal('2017'), Decimal('6'), 101, Decimal('0.00'), Decimal('3.18')),
 (Decimal('2017'), Decimal('7'), 126, Decimal('24.75'), Decimal('3.19')),
 (Decimal('2017'), Decimal('8'), 137, Decimal('8.73'), Decimal('3.23')),
 (Decimal('2017'), Decimal('9'), 150, Decimal('9.49'), Decimal('3.56')),
 (Decimal('2017'), Decimal('10'), 154, Decimal('2.67'), Decimal('3.38')),
 (Decimal('2017'), Decimal('11'), 231, Decimal('50.00'), Decimal('3.11')),
 (Decimal('2017'), Decimal('12'), 178, Decimal('-22.94'), Decimal('3.18')),
 (Decimal('2018'), Decimal('1'), 239, Decimal('34.27'), Decimal('3.34')),
 (Decimal('2018'), Decimal('2'), 262, Decimal('9.62'), Decimal('3.99')),
 (Decimal('2018'), Decimal('3'), 238, Decimal('-9.16'), Decimal('3.35')),
 (Decimal('2018'), Decimal('4'), 223, Decimal('-6.30'), Decimal('3.24')),
 (Decimal('2018'), Decimal('5'), 243, Decimal('8.97'), Decimal('3.57')),
 (Decimal('2018'), Decimal('6'), 221, Decimal('-9.05'), Decimal('3.61')),
 (Decimal('2018'), Decimal('7'), 218, Decimal('-1.36'), Decimal('3.50')),
 (Decimal('2018'), Decimal('8'), 232, Decimal('6.42'), Decimal('3.59')),
 (Decimal('2018'), Decimal('9'), 9, Decimal('-96.12'), Decimal('64.29')),
 (Decimal('2018'), Decimal('10'), 3, Decimal('-66.67'), Decimal('75.00'))]

Breaking the numbers into monthly basic, it clearly shows that there was an increasing trend in the number of repeat customers mostly in 2017. That trend has not lasted long until 2018 when the growth rate fluctuated relatively during the year. 

## Customer analysis

First, we would like to know whether there was any customers who did not purchase during the period

In [7]:
%%sql
SELECT customer_unique_id, COUNT(order_id) AS no_of_orders
FROM customers
    LEFT JOIN orders USING (customer_id)
GROUP BY customer_unique_id
ORDER BY no_of_orders ASC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('0000b849f77a49e4a4ce2b2a4ca5be3f', 1),
 ('0000f46a3911fa3c0805444483337064', 1),
 ('0000f6ccb0745a6a4b88665a16c9f078', 1),
 ('0004aac84e0df4da2b147fca70cf8255', 1),
 ('0004bd2a26a76fe21f786e4fbd80607f', 1),
 ('00050ab1314c0e55a6ca13cf7181fecf', 1),
 ('00053a61a98854899e70ed204dd4bafe', 1),
 ('0005e1862207bf6ccc02e4228effd9a0', 1),
 ('0005ef4cd20d2893f0d9fbd94d3c0d97', 1),
 ('0000366f3b9a7992bf8c76cfdf3221e2', 1)]

As each customer purchased at least one order, we are good to go to find out which cities that mostly our customers are in.

In [95]:
%%sql
SELECT customer_city,
        COUNT(DISTINCT customer_unique_id) AS no_of_customers,
        ROUND(CAST(CAST(COUNT(DISTINCT customer_unique_id) AS REAL) / (SELECT COUNT(DISTINCT customer_unique_id) 
                                                            FROM customers)*100 AS NUMERIC), 2) AS percentage
FROM customers
GROUP BY customer_city
ORDER BY no_of_customers DESC 
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('sao paulo', 14984, Decimal('15.59')),
 ('rio de janeiro', 6620, Decimal('6.89')),
 ('belo horizonte', 2672, Decimal('2.78')),
 ('brasilia', 2069, Decimal('2.15')),
 ('curitiba', 1465, Decimal('1.52')),
 ('campinas', 1398, Decimal('1.45')),
 ('porto alegre', 1326, Decimal('1.38')),
 ('salvador', 1209, Decimal('1.26')),
 ('guarulhos', 1153, Decimal('1.20')),
 ('sao bernardo do campo', 908, Decimal('0.94'))]

Not surprisingly, most of Olist's customers are in Sao Paulo and Rio De Janeiro which are also the 2 biggest cities in Brazil. <br>
In addition to number of customers, we also want to see the Revenue and Average Revenue per Order in each city.

In [106]:
%%sql
SELECT customer_city,
        COUNT(od.order_id) AS no_of_orders,
        ROUND(SUM(oi.price + oi.freight_value)) AS revenue,
        ROUND(SUM(oi.price + oi.freight_value) / COUNT(od.order_id)) AS avg_rev_per_order
FROM customers cs
    JOIN orders od USING (customer_id)
    JOIN order_items oi USING (order_id)
GROUP BY customer_city
ORDER BY revenue DESC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('sao paulo', 17808, 2170219.0, 122.0),
 ('rio de janeiro', 7837, 1154230.0, 147.0),
 ('belo horizonte', 3144, 416733.0, 133.0),
 ('brasilia', 2392, 352305.0, 147.0),
 ('curitiba', 1751, 244740.0, 140.0),
 ('porto alegre', 1612, 224064.0, 139.0),
 ('salvador', 1412, 216772.0, 154.0),
 ('campinas', 1654, 212542.0, 129.0),
 ('guarulhos', 1329, 163576.0, 123.0),
 ('niteroi', 980, 137919.0, 141.0)]

Similarly, these 2 biggest cities also accounted for a large amount of revenue of Olist. However, high revenue and lots of customers do not mean the average revenue per order will be high as well. <br>
Despite of the fact that Sao Paulo had the highest revenue and a large number of orders, the average revenue per order was the lowest among top 10 cities regarding revenue. 

In [250]:
%%sql
SELECT customer_city,
        COUNT(DISTINCT cs.customer_unique_id) AS no_of_customers,
        ROUND(SUM(oi.price + oi.freight_value)) AS revenue,
        ROUND(SUM(oi.price + oi.freight_value) / COUNT(DISTINCT cs.customer_unique_id)) AS avg_rev_per_customer
FROM customers cs
    JOIN orders od USING (customer_id)
    JOIN order_items oi USING (order_id)
GROUP BY customer_city
ORDER BY revenue DESC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('sao paulo', 14865, 2170215.0, 146.0),
 ('rio de janeiro', 6576, 1154234.0, 176.0),
 ('belo horizonte', 2654, 416733.0, 157.0),
 ('brasilia', 2056, 352305.0, 171.0),
 ('curitiba', 1455, 244740.0, 168.0),
 ('porto alegre', 1320, 224064.0, 170.0),
 ('salvador', 1202, 216772.0, 180.0),
 ('campinas', 1386, 212542.0, 153.0),
 ('guarulhos', 1143, 163576.0, 143.0),
 ('niteroi', 807, 137919.0, 171.0)]

Regarding revenue and number of customer, the average revenue per customer in Sao Paulo was relatively low with only $146 in comparison with other cities. 

Going deeper with 2 metrics: Average revenue per order and Average revenue per customer, we would like to see where Sao Paulo and other top cities are in comparison to the Average of all cities. <br>
Taking the above calculating ways for 2 metrics, we will first store them in temporary table and then later on, we will calculate the different rate between the metric of one city and the average of all cities. 

In [255]:
%%sql

WITH tmp AS
(
    SELECT customer_city,
            COUNT(od.order_id) AS no_of_orders,
            COUNT(DISTINCT cs.customer_unique_id) AS no_of_customers,
            ROUND(SUM(oi.price + oi.freight_value)) AS revenue,
            ROUND(SUM(oi.price + oi.freight_value) / COUNT(od.order_id)) AS avg_rev_per_order,
            ROUND(SUM(oi.price + oi.freight_value) / COUNT(DISTINCT cs.customer_unique_id)) AS avg_rev_per_customer
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
    GROUP BY customer_city
    ORDER BY revenue DESC
    
)

SELECT customer_city, revenue, avg_rev_per_order, avg_rev_per_customer,
        ROUND(CAST(CAST(avg_rev_per_order AS REAL) / 
                                    (SELECT SUM(price + freight_value)/COUNT(order_id) 
                                    FROM orders 
                                        JOIN order_items USING (order_id))
                                    - 1 AS NUMERIC),2) * 100 AS avg_order_vs_all,
        ROUND(CAST(CAST(avg_rev_per_customer AS REAL) / 
                                    (SELECT SUM(price + freight_value)/COUNT(DISTINCT customer_unique_id) 
                                    FROM customers 
                                        JOIN orders USING (customer_id)
                                        JOIN order_items USING (order_id))
                                    - 1 AS NUMERIC),2) * 100 AS avg_customer_vs_all
FROM tmp
LIMIT 5

 * postgresql://postgres:***@localhost/olist
5 rows affected.


[('sao paulo', 2170218.0, 122.0, 146.0, Decimal('-13.00'), Decimal('-12.00')),
 ('rio de janeiro', 1154232.0, 147.0, 176.0, Decimal('5.00'), Decimal('6.00')),
 ('belo horizonte', 416733.0, 133.0, 157.0, Decimal('-5.00'), Decimal('-5.00')),
 ('brasilia', 352305.0, 147.0, 171.0, Decimal('5.00'), Decimal('3.00')),
 ('curitiba', 244740.0, 140.0, 168.0, Decimal('0.00'), Decimal('1.00'))]

While there was around +-5% in ratio difference in most of cities in top 5, Sao Paulo had lower rate in both 2 metrics of Average revenue per order and per customer with 13% and 12% lower than the average of all cities, respectively.

We will continuously analyse the location by talking more about the customer's buying behaviour. The question is: Between purchasing in the same city and in other cities, which one would be preferable by customers? <br>
The idea here is to calculate the number of orders overall (assigned as tmp table), the number of orders counted where customers purchased in the same city (assigned as tmp1) and count orders where they were purchased in different cities from the customer's original city. 

In [167]:
%%sql

WITH tmp AS
(
    SELECT cs.customer_city,
            CAST(COUNT(od.order_id) AS REAL) AS no_of_orders
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN sellers sl USING (seller_id)
    GROUP BY cs.customer_city
    
),

tmp1 AS
(
    SELECT cs.customer_city,
            CAST(COUNT(od.order_id) AS REAL) AS same_orders
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN sellers sl USING (seller_id)
    WHERE cs.customer_city = sl.seller_city
    GROUP BY cs.customer_city
    
)

SELECT customer_city, 
        ROUND(CAST(CAST(same_orders AS REAL) / no_of_orders AS NUMERIC),2) AS same_ratio,
        1 - ROUND(CAST(CAST(same_orders AS REAL) / no_of_orders AS NUMERIC),2) AS diff_ratio
FROM tmp
JOIN tmp1 USING (customer_city)
ORDER BY no_of_orders DESC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('sao paulo', Decimal('0.27'), Decimal('0.73')),
 ('rio de janeiro', Decimal('0.04'), Decimal('0.96')),
 ('belo horizonte', Decimal('0.05'), Decimal('0.95')),
 ('brasilia', Decimal('0.02'), Decimal('0.98')),
 ('curitiba', Decimal('0.07'), Decimal('0.93')),
 ('campinas', Decimal('0.02'), Decimal('0.98')),
 ('porto alegre', Decimal('0.02'), Decimal('0.98')),
 ('salvador', Decimal('0.00'), Decimal('1.00')),
 ('guarulhos', Decimal('0.02'), Decimal('0.98')),
 ('sao bernardo do campo', Decimal('0.01'), Decimal('0.99'))]

At Olist, customers purchased mostly in different cities from their original city. <br>
Interestingly, there were still a lof of customers in Sao Paulo purchased goods in that city also with 27%, which is the highest amount in comparison to other cities.   

Next, let's see the difference between the Average revenue per order purchased in the same city and also in other cities. <br>
The idea is to calculate the revenue of orders purchased in the same city with customer's one (assigned as tmp1 table) and the same logic is also applied to tmp2 table, which is about the revenue of orders purchased outside the city of customer. Finally, each of them will be divided to total number of orders in the same city and other cities, respectively.   

In [182]:
%%sql

WITH tmp1 AS
(
    SELECT cs.customer_city,
            ROUND(SUM(oi.price + oi.freight_value)) AS same_revenue
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN sellers sl USING (seller_id)
    WHERE cs.customer_city = sl.seller_city
    GROUP BY cs.customer_city
    
),

tmp2 AS
(
    SELECT cs.customer_city,
            ROUND(SUM(oi.price + oi.freight_value)) AS diff_revenue
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN sellers sl USING (seller_id)
    WHERE cs.customer_city != sl.seller_city
    GROUP BY cs.customer_city
)



SELECT customer_city, 
        ROUND(CAST(CAST(same_revenue AS REAL) / (SELECT COUNT(od.order_id)
                                                 FROM customers cs
                                                     JOIN orders od USING (customer_id)
                                                     JOIN order_items oi USING (order_id)
                                                     JOIN sellers sl USING (seller_id)
                                                 WHERE cs.customer_city = sl.seller_city) 
                                                       AS NUMERIC),2) AS same_avg_order,
        ROUND(CAST(CAST(diff_revenue AS REAL) / (SELECT COUNT(od.order_id)
                                                 FROM customers cs
                                                     JOIN orders od USING (customer_id)
                                                     JOIN order_items oi USING (order_id)
                                                     JOIN sellers sl USING (seller_id)
                                                 WHERE cs.customer_city != sl.seller_city) 
                                                       AS NUMERIC),2) AS diff_avg_order
FROM tmp1 
JOIN tmp2 USING (customer_city)
ORDER BY same_revenue DESC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('sao paulo', Decimal('77.63'), Decimal('16.06')),
 ('rio de janeiro', Decimal('9.47'), Decimal('10.29')),
 ('curitiba', Decimal('2.56'), Decimal('2.15')),
 ('belo horizonte', Decimal('2.37'), Decimal('3.77')),
 ('guarulhos', Decimal('1.13'), Decimal('1.47')),
 ('campinas', Decimal('0.80'), Decimal('1.95')),
 ('brasilia', Decimal('0.73'), Decimal('3.26')),
 ('porto alegre', Decimal('0.65'), Decimal('2.06')),
 ('caxias do sul', Decimal('0.45'), Decimal('0.39')),
 ('mogi das cruzes', Decimal('0.36'), Decimal('0.53'))]

The above table showed an interesting trend that average revenue per order to products purchased outside the original city of customer were not always higher than the ones purchased in the same city although that metric (diff_avg_order) got all the advantages with very high order quantity and revenue. <br>
For example, in Sao Paulo, the average revenue per order purchased in the same city was $77.63, which was much higher than being purchased in other cities. So, what caused this number too big? 

In [208]:
%%sql

SELECT category_translation,
        COUNT(od.order_id) AS no_of_orders,
        ROUND(SUM(oi.price + oi.freight_value)) AS revenue,
        ROUND(CAST(CAST(SUM(oi.price + oi.freight_value) AS REAL) / COUNT(od.order_id) AS NUMERIC),2) AS avg_order
FROM customers cs
    JOIN orders od USING (customer_id)
    JOIN order_items oi USING (order_id)
    JOIN sellers sl USING (seller_id)
    JOIN products pd ON oi.product_id = pd.product_id
    JOIN product_translation tr ON pd.product_category = tr.category
WHERE cs.customer_city = sl.seller_city
    AND cs.customer_city = 'sao paulo'
GROUP BY category_translation
ORDER BY avg_order DESC
LIMIT 10

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('costruction_tools_tools', 1, 1930.0, Decimal('1930.34')),
 ('home_appliances_2', 9, 6083.0, Decimal('675.85')),
 ('dvds_blu_ray', 3, 1628.0, Decimal('542.55')),
 ('furniture_bedroom', 3, 1535.0, Decimal('511.74')),
 ('office_furniture', 1, 499.0, Decimal('498.66')),
 ('small_appliances_home_oven_and_coffee', 4, 1831.0, Decimal('457.64')),
 ('agro_industry_and_commerce', 18, 4295.0, Decimal('238.62')),
 ('costruction_tools_garden', 10, 1546.0, Decimal('154.60')),
 ('cool_stuff', 215, 32374.0, Decimal('150.58')),
 ('signaling_and_security', 13, 1915.0, Decimal('147.31'))]

When we find out about the products that had the highest average revenue per order in Sao Paulo, things like construction tools was only purchased once but their base price was at a very high amount, which eventually caused the average revenue per order of this product category to be on top. Additionally, things in big size with complex structure such as home appliance, bedroom or office furniture were categories which customers in Sao Paulo preferred to purchase within this city.

Lastly, we would like to see the most popular product categories between 2 types of customer: those who purchased products in the same city and those who purchased from seller at a different city from theirs. <br>
The idea here is to find out top 5 most popular categories under "Same City" type and "Different City" type, and they will be combined into one by using UNION ALL.

In [216]:
%%sql

SELECT * FROM
(
    SELECT 'Same City' AS location_type, 
            tr.category_translation,
            COUNT(od.order_id) AS no_of_orders
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN products pd USING (product_id)
        JOIN sellers sl USING (seller_id)
        JOIN product_translation tr ON pd.product_category = tr.category
    WHERE cs.customer_city = sl.seller_city
    GROUP BY location_type, tr.category_translation
    ORDER BY no_of_orders DESC
    LIMIT 5
) a

UNION ALL

SELECT * FROM
(
    SELECT 'Different City' AS location_type, 
            tr.category_translation,
            COUNT(od.order_id) AS no_of_orders
    FROM customers cs
        JOIN orders od USING (customer_id)
        JOIN order_items oi USING (order_id)
        JOIN products pd USING (product_id)
        JOIN sellers sl USING (seller_id)
        JOIN product_translation tr ON pd.product_category = tr.category
    WHERE cs.customer_city != sl.seller_city
    GROUP BY location_type, tr.category_translation
    ORDER BY no_of_orders DESC
    LIMIT 5
) b

 * postgresql://postgres:***@localhost/olist
10 rows affected.


[('Same City', 'health_beauty', 535),
 ('Same City', 'housewares', 482),
 ('Same City', 'furniture_decor', 453),
 ('Same City', 'sports_leisure', 444),
 ('Same City', 'telephony', 409),
 ('Different City', 'bed_bath_table', 10835),
 ('Different City', 'health_beauty', 9135),
 ('Different City', 'sports_leisure', 8197),
 ('Different City', 'furniture_decor', 7881),
 ('Different City', 'computers_accessories', 7503)]

From the above table, health beauty, furniture for decoration and sports leisure were the ones that mostly purchased regardless of purchasing location.

## Conclusion

- The performance of Olist was good in 2017 and 2018 with a high growth rate in both number of overall customers and repeat customers. Though the growth rate of repeat customers was increasing at a faster pace than the rate of overall customers, the number of repeat customers only accounted for a very small amount of all customers Olist had.

- As we could expect, most customers of Olist came from biggest cities of Brazil like Sao Paulo, which was the biggest market of Olist based on revenue. Despite of that title, the average revenue per order and per customers in this city were much lower than its counterpart in Brazil. 
- Customers tend to purchase products from outside their cities. And regardless of location, health beauty, furnitures and sports leisure were the most popular product categories based on the number of orders.